# Loading the Data & Libraries

In [165]:
library(class)

In [166]:
library(car)

In [178]:
 df <- read.csv("german_credit_data1.csv",header = TRUE)
  

head(df)

X. Age Sex    Job Housing Saving.accounts Checking.account Credit.amount
1 0  67  male   2   own     NA              little           1169         
2 1  22  female 2   own     little          moderate         5951         
3 2  49  male   1   own     little          NA               2096         
4 3  45  male   2   free    little          little           7882         
5 4  53  male   2   free    little          little           4870         
6 5  35  male   1   free    NA              NA               9055         
  Duration Purpose             Credit.Risks
1  6       radio/TV            1           
2 48       radio/TV            2           
3 12       education           1           
4 42       furniture/equipment 1           
5 24       car                 2           
6 36       education           1

# Chaging the NA values to 0

In [168]:
df$Saving.accounts<-as.character(df$Saving.accounts)
df$Checking.account<-as.character(df$Checking.account)
df$Saving.accounts[is.na(df$Saving.accounts)]<-0
df$Checking.account[is.na(df$Checking.account)]<-0

In [169]:
nrow(df)

[1] 1000

In [170]:
head(df,50)

X. Age Sex    Job Housing Saving.accounts Checking.account Credit.amount
1   0 67  male   2   own     0               little            1169        
2   1 22  female 2   own     little          moderate          5951        
3   2 49  male   1   own     little          0                 2096        
4   3 45  male   2   free    little          little            7882        
5   4 53  male   2   free    little          little            4870        
6   5 35  male   1   free    0               0                 9055        
7   6 53  male   2   own     quite rich      0                 2835        
8   7 35  male   3   rent    little          moderate          6948        
9   8 61  male   1   own     rich            0                 3059        
10  9 28  male   3   own     little          moderate          5234        
11 10 25  female 2   rent    little          moderate          1295        
12 11 24  female 2   rent    little          little            4308        
13 12 22  female 2   own     little          moderate          1567        
14 13 60  male   1   own     little          little            1199        
15 14 28  female 2   rent    little          little            1403        
16 15 32  female 1   own     moderate        little            1282        
17 16 53  male   2   own     0               0                 2424        
18 17 25  male   2   own     0               little            8072        
19 18 44  female 3   free    little          moderate         12579        
20 19 31  male   2   own     quite rich      0                 3430        
21 20 48  male   2   own     little          0                 2134        
22 21 44  male   2   rent    quite rich      little            2647        
23 22 48  male   1   rent    little          little            2241        
24 23 44  male   2   own     moderate        moderate          1804        
25 24 26  male   2   own     0               0                 2069        
26 25 36  male   1   own     little          little            1374        
27 26 39  male   1   own     little          0                  426        
28 27 42  female 2   rent    rich            rich               409        
29 28 34  male   2   own     little          moderate          2415        
30 29 63  male   2   own     little          little            6836        
31 30 36  male   2   own     rich            moderate          1913        
32 31 27  male   2   own     little          little            4020        
33 32 30  male   2   own     moderate        moderate          5866        
34 33 57  male   1   rent    0               0                 1264        
35 34 33  female 3   own     little          rich              1474        
36 35 25  male   1   own     little          moderate          4746        
37 36 31  male   2   free    little          0                 6110        
38 37 37  male   2   own     little          rich              2100        
39 38 37  male   2   own     little          rich              1225        
40 39 24  male   2   own     little          moderate           458        
41 40 30  male   3   own     quite rich      0                 2333        
42 41 26  male   2   own     quite rich      moderate          1158        
43 42 44  male   1   own     little          moderate          6204        
44 43 24  male   2   rent    moderate        little            6187        
45 44 58  female 1   free    little          little            6143        
46 45 35  female 3   own     little          0                 1393        
47 46 39  male   2   own     quite rich      0                 2299        
48 47 23  female 0   rent    quite rich      little            1352        
49 48 39  male   1   own     little          0                 7228        
50 49 28  female 2   own     moderate        0                 2073        
   Duration Purpose             Credit.Risks
1   6       radio/TV            1           
2  48       radio/TV            2    

In [171]:
summary(df)

       X.             Age            Sex           Job        Housing   
 Min.   :  0.0   Min.   :19.00   female:310   Min.   :0.000   free:108  
 1st Qu.:249.8   1st Qu.:27.00   male  :690   1st Qu.:2.000   own :713  
 Median :499.5   Median :33.00                Median :2.000   rent:179  
 Mean   :499.5   Mean   :35.55                Mean   :1.904             
 3rd Qu.:749.2   3rd Qu.:42.00                3rd Qu.:2.000             
 Max.   :999.0   Max.   :75.00                Max.   :3.000             
                                                                        
 Saving.accounts    Checking.account   Credit.amount      Duration   
 Length:1000        Length:1000        Min.   :  250   Min.   : 4.0  
 Class :character   Class :character   1st Qu.: 1366   1st Qu.:12.0  
 Mode  :character   Mode  :character   Median : 2320   Median :18.0  
                                       Mean   : 3271   Mean   :20.9  
                                       3rd Qu.: 3972   3rd Qu.:24.

# Changing the text in Savings and Checking account to numerical data


In [172]:

df$Saving.accounts<-ifelse(df$Saving.accounts == "little", 1, ifelse(df$Saving.accounts== "moderate", 2,ifelse(df$Saving.accounts == "quite rich",3,ifelse(df$Saving.accounts == "rich",4,0))))

df$Checking.account<-ifelse(df$Checking.account == "little", 1, ifelse(df$Checking.account== "moderate", 2,ifelse(df$Checking.account == "rich",3,0)))



# Normalizing the data for accurate distance calculation

In [173]:

df$Age<-((df$Age - min(df$Age)) / (max(df$Age) - min(df$Age)))

df$Saving.accounts<-((df$Saving.accounts - min(df$Saving.accounts)) / (max(df$Saving.accounts) - min(df$Saving.accounts)))

df$Checking.account<-((df$Checking.account - min(df$Checking.account)) / (max(df$Checking.account) - min(df$Checking.account)))
      
#df$Credit.Risks<-((df$Credit.Risks - min(df$Credit.Risks)) / (max(df$Credit.Risks) - min(df$Credit.Risks)))    

# Dataframe with normalized Values

In [174]:
df

X.  Age        Sex    Job Housing Saving.accounts Checking.account
1     0  0.85714286 male   2   own     0.00            0.3333333       
2     1  0.05357143 female 2   own     0.25            0.6666667       
3     2  0.53571429 male   1   own     0.25            0.0000000       
4     3  0.46428571 male   2   free    0.25            0.3333333       
5     4  0.60714286 male   2   free    0.25            0.3333333       
6     5  0.28571429 male   1   free    0.00            0.0000000       
7     6  0.60714286 male   2   own     0.75            0.0000000       
8     7  0.28571429 male   3   rent    0.25            0.6666667       
9     8  0.75000000 male   1   own     1.00            0.0000000       
10    9  0.16071429 male   3   own     0.25            0.6666667       
11   10  0.10714286 female 2   rent    0.25            0.6666667       
12   11  0.08928571 female 2   rent    0.25            0.3333333       
13   12  0.05357143 female 2   own     0.25            0.6666667       
14   13  0.73214286 male   1   own     0.25            0.3333333       
15   14  0.16071429 female 2   rent    0.25            0.3333333       
16   15  0.23214286 female 1   own     0.50            0.3333333       
17   16  0.60714286 male   2   own     0.00            0.0000000       
18   17  0.10714286 male   2   own     0.00            0.3333333       
19   18  0.44642857 female 3   free    0.25            0.6666667       
20   19  0.21428571 male   2   own     0.75            0.0000000       
21   20  0.51785714 male   2   own     0.25            0.0000000       
22   21  0.44642857 male   2   rent    0.75            0.3333333       
23   22  0.51785714 male   1   rent    0.25            0.3333333       
24   23  0.44642857 male   2   own     0.50            0.6666667       
25   24  0.12500000 male   2   own     0.00            0.0000000       
26   25  0.30357143 male   1   own     0.25            0.3333333       
27   26  0.35714286 male   1   own     0.25            0.0000000       
28   27  0.41071429 female 2   rent    1.00            1.0000000       
29   28  0.26785714 male   2   own     0.25            0.6666667       
30   29  0.78571429 male   2   own     0.25            0.3333333       
⋮    ⋮   ⋮          ⋮      ⋮   ⋮       ⋮               ⋮               
971  970 0.05357143 male   2   own     0.50            0.6666667       
972  971 0.42857143 male   1   own     0.25            0.0000000       
973  972 0.17857143 female 0   rent    0.25            0.3333333       
974  973 0.30357143 male   2   rent    0.25            0.3333333       
975  974 0.25000000 female 2   own     0.25            0.0000000       
976  975 0.67857143 female 1   own     0.75            1.0000000       
977  976 0.80357143 female 2   own     0.25            0.6666667       
978  977 0.41071429 male   2   own     0.00            0.6666667       
979  978 0.50000000 male   1   own     0.25            0.0000000       
980  979 0.10714286 male   2   rent    0.50            0.6666667       
981  980 0.53571429 male   2   own     0.25            0.6666667       
982  981 0.25000000 male   3   rent    0.25            0.0000000       
983  982 0.16071429 female 3   own     0.50            1.0000000       
984  983 0.12500000 male   2   own     0.25            0.3333333       
985  984 0.19642857 male   1   own     0.25            0.0000000       
986  985 0.10714286 female 2   rent    0.25            0.3333333       
987  986 0.25000000 male   2   own     0.25            1.0000000       
988  987 0.80357143 female 2   own     0.50            0.0000000       
989  988 0.17857143 male   3   free    0.25            0.3333333       
990  989 0.51785714 male   1   own     0.25            0.6666667       
991  990 0.32142857 male   1   own     0.00            0.0000000       
992  991 0.26785714 male   1   own     0.50            0.0000000       
993  992 0.07142857 male   1   rent    0.00            0.3333333       
994  993 0.19642857 male   3   own     0.25  

# KNN Function

### Passing: Age, Savings Account , Checking Account 
### Target: Credit Risk

In [175]:
knncalc<-function (df,sampling.rate){

n.points <- nrow(df) # number of rows in the dataset
num.test.set.labels <- n.points * (1 - sampling.rate)
# randomly sample which rows will go in the training set
training <- sample(1:n.points, sampling.rate * n.points,
replace=FALSE)
train <- subset(df[training, ], select = c(Age,Saving.accounts,Checking.account))

testing <- setdiff(1:n.points, training)
# define the test set to be the other rows
test <- subset(df[testing, ], select = c(Age,Saving.accounts,Checking.account))
cl <- df$Credit.Risks[training]
# this is the subset of labels for the training set
true.labels <- df$Credit.Risks[testing]

d2 = data.frame( k=rep(0, 20), misclassification.rate=rep(0,20))

for (k in 1:20) {
#print(k)
predicted.labels <- knn(train, test, cl, k)
# We're using the R function knn()
num.incorrect.labels <- sum(predicted.labels != true.labels)
misclassification.rate <- num.incorrect.labels /num.test.set.labels
    d2[k,]=c(k,misclassification.rate)
#print(misclassification.rate)
}
    return (d2)
    }

# Calling th knn Function with sampling rate as the parameter

In [176]:
df0.9<-knncalc(df,0.9)
df0.7<-knncalc(df,0.7)
df0.6<-knncalc(df,0.6)

# Ananlysis of KNN for k varying from 1 to 20 

In [177]:
final<-merge(merge(df0.6,df0.7,by="k"),df0.9,by="k")

colnames(final)<-c("k","misclassification.rate train=60","misclassification.rate train=70","misclassification.rate train=90")

final

k  misclassification.rate train=60 misclassification.rate train=70
1   1 0.3275                          0.3300000                      
2   2 0.3325                          0.3266667                      
3   3 0.3425                          0.3166667                      
4   4 0.3275                          0.3466667                      
5   5 0.3325                          0.3133333                      
6   6 0.3250                          0.3200000                      
7   7 0.3000                          0.3333333                      
8   8 0.2975                          0.3233333                      
9   9 0.3025                          0.3166667                      
10 10 0.2950                          0.3233333                      
11 11 0.2975                          0.3133333                      
12 12 0.2950                          0.3166667                      
13 13 0.3025                          0.3133333                      
14 14 0.2900                          0.3200000                      
15 15 0.2925                          0.3166667                      
16 16 0.2950                          0.3066667                      
17 17 0.2925                          0.3200000                      
18 18 0.2975                          0.3166667                      
19 19 0.3175                          0.3266667                      
20 20 0.3125                          0.3100000                      
   misclassification.rate train=90
1  0.35                           
2  0.31                           
3  0.28                           
4  0.28                           
5  0.26                           
6  0.30                           
7  0.31                           
8  0.29                           
9  0.29                           
10 0.35                           
11 0.37                           
12 0.36                           
13 0.35                           
14 0.38                           
15 0.38                           
16 0.38                           
17 0.38                           
18 0.39                           
19 0.37                           
20 0.39

# Minimum Misclassification Rate=0.2875 at k=3,4 when trainset is 60 %

# Minimum Misclassification Rate=0.28 at k=5 when trainset is 70 %

# Minimum Misclassification Rate=0.22 at k=18 when trainset is 90 %

